crates/conjure_core/src/rule_engine/rewrite_naive.rs

```diff
@@ -0,0 +1,101 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use uniplate::Biplate;
+
+use super::{RewriteError, RuleSet};
+use crate::{
+    ast::Expression as Expr,
+    rule_engine::{get_rule_priorities, get_rules_vec, Reduction},
+    Model,
+};
+
+/// A naive, exhaustive rewriter.
+///
+/// **This rewriter should not be used in production, and is intended as a development tool.**
+///
+/// The goal of this rewriter is to model the correct rule application order. To this end, it uses
+/// the simplest implementation possible, disregarding performance.
+///
+/// **Rule application order:** apply the highest priority rule possible anywhere in the tree,
+/// favouring larger expressions as a tie-breaker.
+
+pub fn rewrite_naive<'a>(
+    model: &Model,
+    rule_sets: &Vec<&'a RuleSet<'a>>,
+) -> Result<Model, RewriteError> {
+    // At each iteration, all rules are checked against all expressions. This is done until no more
+    // rules can be applied to any expression.
+
+    // rules sorted by priority
+
+    let rules = get_rules_vec(&get_rule_priorities(rule_sets).unwrap());
+
+    type CtxFn = Arc<dyn Fn(Expr) -> Vec<Expr>>;
+
+    // store, by priority, a reduction, rule name, expression, and context function to rebuild the
+    // constraints list given a replacement expression
+    let mut applicable_reductions: BTreeMap<u16, (Reduction, String, Expr, CtxFn)> =
+        BTreeMap::new();
+
+    let mut model = model.clone();
+
+    loop {
+        applicable_reductions.clear();
+
+        for (expr, ctx) in Biplate::<Expr>::contexts_bi(&model.get_constraints_vec()) {
```

Comment:
instead: sort rules by priority groups. group one is all of the highest, equal priority rules and so on.

loop over rule groups first.

inside, do the context_bi thing and loop over all expressions

the first applicable rule is the one you want to apply.

if multiple rules are applicable to the same expression, rule selection logic (first / random / ask the user / for now it shouldn't really happen, so check and complain if it does)

should be faster. and you call it naive, but adding clean-dirty to this would go quite a long way towards what we want...